### DISCUSSION TITLE
Abfrage/Nutzung "Netzwerk Gesamtleistung" Victron für Solar-Passthrough

### DISCUSSION THREAD

**Discussion by tonibbb** [hoylabs/OpenDTU-OnBattery#714] [General]
Hallo,

vielen Dank für dieses tolle Projekt und dessen kontinuierliche Weiterentwicklung.
Kurz mein Setup:
- 13 Module, jedes an einem eigen kleinen Victron 75/15 (ja, ich hatte Gründe dafür, könnte ich gerne separat erläutern), die alle
- einen 24 V Speicher, ca. 10 kWh (Eigenbau, Bleiakkus die übrig waren) laden, an dem 
 a) ein HM600 hängt, der über eine openDTU-onBattery gesteuert in eine "moderne Messeinrichtung" einspeist (bzw. versucht einer Nulleinspeisung/maximalen Selbstnutzung nahezukommen)
 b) ein weiterer HM600 mit ebenfalls eigener openDTU-onBattery hängt, der auf das Einstecken des Beretti-Steckers wartet und dann ein anderes Netz im Haus speisen würde, welches sich derzeit noch hinter einem nicht rücklaufgesicherten Ferraris-Zähler befindet 

(Natürlich warte ich mit dem Einstecken und Anmelden des 2. HM600 noch bis Mai und hoffe, dass der Netzbetreiber die gesetzlich vorgeschrieben vollen 4 Monate für einen Tausch des Zählers benötigen wird und ich somit möglichst viele kWh an meinen Stromanbieter "verkaufen" kann; bin mal gespannt ob/wie das funktioniert).

Frage/Problem:
Ich erwarte eine hohe bereitstehende Leistung solange die Sonne scheint und solange der Speicher voll ist. Solange soll der 2. HM600 ordentlich, also mit 600 W einspeisen. Andererseits soll auch für den Rest des Tages/der Nacht ausreichend Ladung im Speicher verbleiben, dass der Bedarf des 1. HM600 gedeckt ist. Soweit ich es verstanden habe, könnte hier Solar-Passthrough die Lösung sein und dass hierfür die DTU über das (bereits installierte) VE.Direct-Kabel von (einem) Victron Laderegler die (Solar-)Leistung abfragt. 
Problem: Die abgefragte Leistung eines Ladereglers widerspiegelt die Gesamtleistung ja nur marginal.
Mögliche Lösung: Die Victrons kommunizieren untereinander und sind in einem "Netzwerk" zusammengeschaltet (funktioniert anstandslos). In der Victron App wird auch eine "Netzwerk Gesamtleistung" angezeigt die aber in openDTU-onBattery nicht angezeigt wird (hier wird nur die Leistung des einen angeschlossenen Ladereglers erfasst). Wenn ich die Gesamtleistung erfassen/nutzen kann, dann sollte auch die Nutzung von Solar-Passthrough gut funktionieren.
Frage:
Wo muss ich ansetzen um die "Netzwerk Gesamtleistung" auszulesen? Und wohin müsste ich die übergeben? Bin leider Laie was Programmieren angeht, kann mir aber vorstellen irgendwelche Werte zu überschreiben (oder sowas ähnliches). Es würde für mich auch reichen, wenn der Wert den sich Solar-Passthrough holt, "Leistung" heißt, solange ich diesen Wert mit dem Wert der "Netzwerk Gesamtleistung" ersetzen kann.
Danke für eure Hinweise.

**Comment by schlimmchen** [comment]
Die aktuelle Implementierung des VE.Direct Protokolls beherrscht nur das Lesen der spontan und regelmäßig von den Ladereglern ausgespuckten Datenpaketen im Text-Format. Darin ist insbesondere nicht die Netzwerk-Gesamtleistung enthalten.

Wonach du fragst ist das aktive Abfrage eines bestimmten Wertes über die HEX-Variante des Protokolls und das anschließende Verarbeiten der Antwort im HEX-Format. Und dann natürlich das Durchreichen dieses Wertes zum Dynamic Power Limiter. Da ist einiges zu tun bis das funktioniert.

> Die abgefragte Leistung eines Ladereglers widerspiegelt die Gesamtleistung ja nur marginal.

Ist das so? Sind die Panele nicht uniform ausgerichtet? Falls doch, könntest du durchaus den vom VE.Direct Protokoll abgeholten Wert für die Leistung mit 13 multiplizieren.

In `src/VictronMppt.cpp` Zeile 97ff:

```
int32_t VictronMpptClass::getPowerOutputWatts() const
{
    int32_t sum = 0;

    for (const auto& upController : _controllers) {
        sum += upController->getData()->P;
    }

    return 13 * sum; // <- hier ist die einzige Anpassung
}
```

**Reply by tonibbb** [reply]
Danke für den Tipp und das raussuchen des Programmabschnitts. Ja, das Multiplizieren scheint mir eine hinreichend genaue Möglichkeit zu sein, auch wenn ich hier mit vielen Verschattungen leben muss - deshalb hat ja jedes Modul seinen eigen Laderegler). Das probiere ich gerne mal aus.

**Reply by SW-Niko** [reply]
Hallo @tonibbb ,
Ich habe da eine Teillösung für dich.  Zu mindestens an die Gesamtleistung komme ich ran. Siehe: #721

**Reply by tonibbb** [reply]
Hallo SW-Niko,
danke, das klingt gut. Ich hätte Interesse an deinem Code und ggf. einer kurzen Anleitung wie dieser zu nutzen ist bzw. wo ich den "einfügen" müsste.

**Comment by SW-Niko** [comment]
Hallo @tonibbb,
Bitte noch etwas Geduld. Die Gesamtleistung bekomme ich schon alle 2 Sekunden vom MPPT. (Siehe Debug Ausgabe)  

10:53:00.597 > [VE.Direct] debug: disassembleHexData(), cmd: 10, id: 0x2027, value: **0x68B0**, Flag: 0x00
10:53:00.645 > [VE.Direct] debug: hexDataHandler(), cmd: 10, id: 0x2027, flag: 0x00, value: **268.00W**, text: 

Die Hex-Schnittstelle mach ich noch fertig und dann muss ich mich in "git" noch etwas einarbeiten um einen "Pull request" machen zu können.
Ich denke an der MPPT Gesamtleistung und erst recht am Hex-Protokoll sind mehrere Leute interessiert und wollen/können sie für weitere Aktionen verwenden.

**Comment by Manos1966** [comment]
Hallo @tonibbb 

> ... wenn ich hier mit vielen Verschattungen leben muss - deshalb hat ja jedes Modul seinen eigen Laderegler

zur Info:
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Victron-MPPT-battery-charger#east--west--panels-with-only-one-victron
